numba/debuginfo.py

```diff
@@ -0,0 +1,378 @@
+        """Util to create constant int in metadata
+        """
+        return ir.IntType(bits)(num)
+
+    @classmethod
+    def _const_bool(cls, boolean):
+        """Util to create constant boolean in metadata
+        """
+        return ir.IntType(1)(boolean)
+
+    #
+    # Helpers to emit the metadata nodes
+    #
+
+    def _di_file(self):
+        return self.module.add_debug_info('DIFile', {
+            'directory': os.path.dirname(self.filepath),
+            'filename': os.path.basename(self.filepath),
+        })
+
+    def _di_compile_unit(self):
+        return self.module.add_debug_info('DICompileUnit', {
+            'language': ir.DIToken('DW_LANG_Python'),
+            'file': self.difile,
+            'producer': 'Numba',
+            'runtimeVersion': 0,
+            'isOptimized': True,
+            'emissionKind': 1,  # 0-NoDebug, 1-FullDebug
+        }, is_distinct=True)
+
+    def _di_subroutine_type(self):
+        return self.module.add_debug_info('DISubroutineType', {
+            'types': self.module.add_metadata([]),
+        })
+
+    def _di_subprogram(self, name, linkagename, line):
+        return self.module.add_debug_info('DISubprogram', {
+            'name': name,
+            'linkageName': linkagename,
+            'scope': self.difile,
+            'file': self.difile,
+            'line': line,
+            'type': self._di_subroutine_type(),
+            'isLocal': False,
+            'isDefinition': True,
+            'scopeLine': line,
+            'isOptimized': True,
+            'variables': self.module.add_metadata([]),
+            'unit': self.dicompileunit,
+        }, is_distinct=True)
+
+    def _di_location(self, line):
+        return self.module.add_debug_info('DILocation', {
+            'line': line,
+            'column': 1,
+            'scope': self.subprograms[-1],
+        })
+
+
+class NvvmDIBuilder(DIBuilder):
+    """
+    Only implemented the minimal metadata to get line number information.
+    See http://llvm.org/releases/3.4/docs/LangRef.html
+    """
+    # These constants are copied from llvm3.4
+    DW_LANG_Python = 0x0014
+    DI_Compile_unit = 786449
+    DI_Subroutine_type = 786453
+    DI_Subprogram = 786478
+    DI_File = 786473
+
+    DWARF_VERSION = None  # don't emit DWARF version
+    DEBUG_INFO_VERSION = 1  # as required by NVVM IR Spec
+    # Rename DIComputeUnit MD to hide it from llvm.parse_assembly()
+    # which strips invalid/outdated debug metadata
+    DBG_CU_NAME = 'numba.llvm.dbg.cu'
+
+    # Default member
+    # Used in mark_location to remember last lineno to avoid duplication
+    _last_lineno = None
+
+    def mark_variable(self, builder, allocavalue, name, lltype, size, loc):
+        # unsupported
+        pass
+
+    def mark_location(self, builder, loc):
+        # Avoid duplication
+        if self._last_lineno == loc.line:
+            return
+        self._last_lineno = loc.line
+        # Add call to an inline asm to mark line location
```

Comment:
Heh, necessary evil to workaround incompatibility between different LLVM version.